design/20220118.server-side-apply.md

```diff
@@ -0,0 +1,167 @@
+status fields on that resource.
+
+## Field Manager and User Agent
+
+The Field Manager is a string denoting the manager who owns a field or set of
+fields on a resource. This string is set by the client at API call time and is
+derived by either the client setting the `fieldManager` URL query explicitly, or
+calculated from the client's user agent ([characters preceding the first "/",
+quote unprintable characters and then trim what's beyond the 128
+limit](https://github.com/kubernetes/kubernetes/blob/9a75e7b0fd1b567f774a3373be640e19b33e7ef1/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go#L252)).
+
+cert-manager will use a consistent naming scheme for both the user agent prefix
+and field manager across all components. Each component will have a human
+readable name that describes that component, and is used for both its user agent
+and field manager. Using the same name enables better auditing trails that helps
+debugging and improves telemetry.
+
+### Scheme
+
+### Field Manager
+
+The field manager is derived from the component in cert-manager, prefixed by the
+string `cert-manager`:
+
+```text
+cert-manager-<cert-manager component>
+```
+
+e.g.
+
+```text
+cert-manager-certificates-issuing
+```
+
+### User Agent
+
+The user agent is inspired by the [client-go default user
+agent](https://github.com/kubernetes/client-go/blob/664b1a6c8ce9c92ce65bef3f9833b402449c98d2/rest/config.go#L499),
+and is defined with the following:
+
+```text
+<component field manager>/<cert-manager version> (<Operating System>/<Architecture>) cert-manager/<git commit>
+```
+
+e.g:
+
+```text
+cert-manager-certificates-issuing/v1.7.0 (Linux/amd64) cert-manager/0b686b8f38c8c7442744c9224d18e780ee7f244a
+```
+
+## Implementation Consideration
+
+All CREATE and UPDATE operations in cert-manager controllers are to be replaced
+with APPLY. For this change, controllers will need to be modified so that for
+these calls, _only_ the fields in which they manage should be included in the
+PATCH API call. This means each controller will need to define exactly which
+fields they are concerned with.
+
+Some fields, such as the Certificate Issuing Condition are managed by more than
+one controller (issuing and trigger Certificate controllers), and as such, will
+need to make use of the `force` option in their API calls. This option tells the
+API server to revoke management of that field from the previous owner, overwrite
+the field, and change owner ship to the new client.
+
+The [fake client-go client](https://github.com/kubernetes/client-go/issues/970)
+does not support the Apply patch call for mocking API calls and events. This
+means that significant controller unit-testing will either need to moved to
+testing against a real API server as integration tests, the controller
+[test framework must add custom support for Apply](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/test/context_builder.go),
+or a new testing framework should be developed.
```

Comment:
Can we implement this upstream?

Reply:
Yep, we can certainly do that. I have added a note that a stop gap would always be required though (perhaps a fork) while we wait for a release.